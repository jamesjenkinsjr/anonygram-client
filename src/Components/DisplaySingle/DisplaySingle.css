.DisplaySingle {
  margin-top: 3em;
}