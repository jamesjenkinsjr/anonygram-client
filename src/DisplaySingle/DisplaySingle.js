import React from 'react';
import CommentFeed from '../CommentFeed/CommentFeed';

export default function DisplaySingle (props) {
  const { image_url, description, timestamp, karma_total, comments} = props.photoObj
  return (
    <div>
      <img src={image_url} alt={description}/>
      <div>{timestamp}</div>
      <div>{karma_total}</div>
      <p>{description}</p>
      <CommentFeed comments={comments}/>
    </div>
  )
};