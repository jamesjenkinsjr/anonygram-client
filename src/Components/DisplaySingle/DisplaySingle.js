import React, { Component } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import CommentFeed from '../CommentFeed/CommentFeed';
import CommentApi from '../../services/comment-api-service';
import ImageContext from '../../contexts/ImageContext';
import TokenService from '../../services/token-service';
import { KeyboardArrowUp } from '@material-ui/icons';
import './DisplaySingle.css';

export default class DisplaySingle extends Component {
  static contextType = ImageContext; 

  state = {
    image: {}, 
    comments: [], 
    loading: false, 
  }

  convertTime = (timestamp) => {
    return moment((new Date(timestamp)).toString()).fromNow();
  }

  componentDidMount = () => {
    const submissionId = this.props.submissionId;
    const image = this.context.images.find(img => img.id === Number(submissionId));

    this.setState({ loading: true });
    CommentApi.getComments(submissionId)
      .then((res) => {
        this.setState({
          image,
          comments: res,
          loading: false,
        })
      });
  }

  setCommentsByPush = (comment) => {
    const commentsArr = this.state.comments.map(c => c);
    commentsArr.push(comment);
    this.setState({ comments: commentsArr })
  } 

  render = () => {
    const userLoggedIn = TokenService.hasAuthToken();

    //This conditional is needed to keep componentDidMount in CommentFeed from running before the needed props are ready to be passed in (to generate usernames)
    if (this.state.comments == null) {
      return null;
    }
    if (this.state.loading === true) {
      return <div className="loader"></div>;
    } else {
      const { id, image_url, image_text, create_timestamp, karma_total } = this.state.image
      return (
        <div className='DisplaySingle'>
          <img src={image_url} alt={image_text}/>
          { /* If user is logged in, render enabled upvote button */
          userLoggedIn ? (
            <div className='DisplaySingle__div upvoteButton' onClick={() => this.context.handleUpvote(id)}>
              <KeyboardArrowUp fontSize="large"/>
              {karma_total}
            </div>
          ) : (
            <div>{karma_total}</div>
          )}
          <div>{this.convertTime(create_timestamp)}</div>

          <p>{image_text}</p>
          <CommentFeed id={this.state.image.id} comments={this.state.comments} setCommentsByPush={this.setCommentsByPush} userLoggedIn={userLoggedIn}/>
        </div>
      )
    }
  }
}