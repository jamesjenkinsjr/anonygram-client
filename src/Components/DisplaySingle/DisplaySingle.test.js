import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import DisplaySingle from './DisplaySingle';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('DisplaySingle component', () => {
  DisplaySingle.contextTypes = {
    images: PropTypes.array,
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<DisplaySingle />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  const props = {
    submissionId: 1,
  };
  const context = {
    images: [
      {
        create_timestamp: new Date('2019-12-10T00:40:55.725Z'),
        id: 1,
        image_url: 'https://via.placeholder.com/150',
        image_text: 'this is image text',
        karma_total: 3,
      },
    ],
    incrementUpvotes: () => {}
  };

  const state = {
    image: {
      create_timestamp: new Date('2019-12-10T00:40:55.725Z'),
      id: 1,
      image_url: 'https://via.placeholder.com/150',
      image_text: 'this is image text',
      karma_total: 3,
    },
    comments: [
      {
        comment_timestamp: new Date('2019-12-10T00:40:55.725Z'),
        user_id: '2d644486-26dc-4ad3-b460-3c6895aa260f',
        comment_id: '2d644486-26dc-4ad3-b460-3c6895aa260f',
        comment_text: 'some comment text',
      },
    ],
  };
  it('renders a submission given props and context', async () => {
    const wrapper = shallow(<DisplaySingle {...props} />, { context });
    wrapper.setState({
      loading: false,
      image: state.image,
      comments: state.comments,
    });
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
