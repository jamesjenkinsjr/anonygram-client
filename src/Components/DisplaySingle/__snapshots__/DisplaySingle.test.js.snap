// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplaySingle component renders a submission given props and context 1`] = `
<div
  className="DisplaySingle"
>
  <img
    alt="this is image text"
    src="https://via.placeholder.com/150"
  />
  <div
    className="DisplaySingle__div upvoteButton"
    onClick={[Function]}
  >
    <KeyboardArrowUpIcon
      fontSize="large"
    />
    3
  </div>
  <div>
    16 hours ago
  </div>
  <p>
    this is image text
  </p>
  <CommentFeed
    comments={
      Array [
        Object {
          "comment_id": "2d644486-26dc-4ad3-b460-3c6895aa260f",
          "comment_text": "some comment text",
          "comment_timestamp": 2019-12-10T00:40:55.725Z,
          "user_id": "2d644486-26dc-4ad3-b460-3c6895aa260f",
        },
      ]
    }
    id={1}
    setCommentsByPush={[Function]}
  />
</div>
`;
